### PR TITLE
Fix main prover in parallel_advance_proof not being closed

### PR DESCRIPTION
**Pull request by nwatson22**
`main_prover`, the `Prover` used by the main thread in `parallel_advance_proof` was not being closed. This fixes it by wrapping the proof with a `with create_prover() as main_prover:` block.

### CODE REVIEW

**Comment by ehildenb** [comment]
Can we test this?

**Comment by nwatson22** [comment]
> Can we test this?

@ehildenb I added a test to check `close()` is called the same number of times a Prover is created.

File: pyk/src/pyk/proof/proof.py [hunk 1]
```diff
@@ -344,43 +344,42 @@ def parallel_advance_proof(
     explored: set[PS] = set()
     iterations = 0
 
-    main_prover = create_prover()
+    with create_prover() as main_prover:
+        main_prover.init_proof(proof)
 
-    main_prover.init_proof(proof)
+        with _ProverPool[P, PS, SR](create_prover=create_prover, max_workers=max_workers) as pool:
```
Comment (tothtamas28) [inline pyk/src/pyk/proof/proof.py:350]: What's the relationship between `main_prover` and the provers created with `create_prover`?

Comment (nwatson22) [reply pyk/src/pyk/proof/proof.py:350]: `main_prover` is also created with `create_prover`, but the difference is that the other provers are created by `_ProverPool` as needed for each thread, but `main_prover` is created directly in the proving loop and is used on the main thread to `init_proof` and compute failure info.

Comment (tothtamas28) [reply pyk/src/pyk/proof/proof.py:350]: Do the other provers also call  `init_proof` and compute `failure_info`? If not, these are two disjoint roles, and the interfaces can be segregated.

**Review by tothtamas28 — APPROVED** [review]

File: pyk/src/pyk/proof/proof.py (hunk 1 above)
Comment (nwatson22) [reply pyk/src/pyk/proof/proof.py:350]: They do not. Really the only thing that's shared is the `kcfg_explore` is used for both. (and `advance_proof`, the older sequential proving loop, is a member of `Proof`, so it uses `self` for all of these functions. I've opened an issue for this at #4333.

**Pull request merged**